### PR TITLE
Rebeka: Code review

### PR DESCRIPTION
**Pull request by merelkuijs**

### CODE REVIEW

File: DM2C/code/model.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+import logging
+import os
+import itertools
+import pdb
+
+import torch
+import torch.optim as optim
+from torch.utils.data import DataLoader
+from tensorboardX import SummaryWriter
+from utils import *#, SFeatDataSet
```
Comment (merelkuijs) [inline DM2C/code/model.py]: I prefer to just import the classes and functions I need.

File: DM2C/code/model.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                decoder_layers.append(nn.BatchNorm1d(self.channels[i - 1]))
+        self.decoder = nn.Sequential(*decoder_layers)
+
+    def forward(self, x):
+        latent = self.encoder(x)
+        output = self.decoder(latent)
+        return output, latent
+
+class CellPLM_model():
+    def __init__(self, model, device='cpu'):
```
Comment (merelkuijs) [inline DM2C/code/model.py]: It would be good to add a bit of documentation. Already specifying the type (self, model: str, device: str = 'cpu') is pretty helpful.

File: DM2C/code/model.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        self.pipeline = CellEmbeddingPipeline(pretrain_prefix=model,  # Specify the pretrain checkpoint to load
+                                         pretrain_directory='../../../models/cellplm/')
+        self.device = device
+    def calc_embed(self, data):
```
Comment (merelkuijs) [inline DM2C/code/model.py]: You could also add the type here: (self, data: anndata.AnnData). Make sure to import anndata at the start of the file if you do this.

File: DM2C/code/model.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        self.decoder = nn.Sequential(*decoder_layers)
+
+    def forward(self, x):
+        latent = self.encoder(x)
+        output = self.decoder(latent)
+        return output, latent
+
+class CellPLM_model():
+    def __init__(self, model, device='cpu'):
+        self.pipeline = CellEmbeddingPipeline(pretrain_prefix=model,  # Specify the pretrain checkpoint to load
```
Comment (merelkuijs) [inline DM2C/code/model.py:63]: Where is the models/cellplm folder? I don't think I was able to find it in your repo.

File: DM2C/code/model.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+        for k, v in sorted(self.config.items()):
+            self.logger.debug("{0}: {1}".format(k, v))
+
+        # Encoders
+        self.cell_plm = CellPLM_model(self.args.cellplm_model)
+        self.vis_trans = Vision_Trans(self.args.hugging_face)
+
+        # Generator
+        # adjusted for the masterpraktikum
+        self.latent_dim_img = self.config['img_input_dim']
```
Comment (merelkuijs) [inline DM2C/code/model.py]: I would advise to change the name of this parameter in self.config. I think img_latent_dim is more suitable given we work with image embeddings rather than the raw input images.

File: DM2C/code/model.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        self.latent_dim_txt = self.config['txt_input_dim']
```
Comment (merelkuijs) [inline DM2C/code/model.py]: Similarly, I would change txt_input_dim to txt_latent_dim.

File: DM2C/code/model.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+class Vision_Trans():
+        def __init__(self, hugging_face):
+            self.processor = AutoImageProcessor.from_pretrained(hugging_face)
+            self.model = AutoModel.from_pretrained(hugging_face)
+
+        def calc_embed(self, imgs):
+            def infer(img):
+                img = Image.open(img).convert("RGB")
+                inputs = self.processor(img, return_tensors="pt")  # preprocesses for correct input format
+                outputs = self.model(**inputs)
```
Comment (merelkuijs) [inline DM2C/code/model.py]: Consider wrapping line 76 as follows:
```
with torch.no_grad():
    outputs = self.model(**inputs)
```
    
Then, your model won't have to compute gradients, which will hopefully speed up image processing. It is not necessary to compute gradients here, because we are not training the model yet.

File: DM2C/code/model.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+                return outputs.pooler_output.detach().squeeze(dim=0)
```
Comment (merelkuijs) [inline DM2C/code/model.py]: If you wrap `outputs` as shown above, you won't have to call `detach()`, I think.

File: DM2C/code/model.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+                weight_decay=self.args.weight_decay)
+            self.optimizer_G = optim.Adam(
+                params,
+                lr=self.args.lr_g, betas=(self.args.b1, self.args.b2),
+                weight_decay=self.args.weight_decay)
+
+        self.set_writer()
+        self.adv_loss_fn = F.binary_cross_entropy_with_logits
+
+    # no pretraining needed (def deleted)
```
Comment (merelkuijs) [inline DM2C/code/model.py:164]: Good!

File: DM2C/code/model.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    def train(self, epoch):
+        self.set_model_status(training=True)
+        for step, (txt_embed, img_embed) in enumerate(self.train_loader):
+            # -----------------
+            #  Train Generator
```
Comment (merelkuijs) [inline DM2C/code/model.py:169]: This section looks correct to me, well done!

File: DM2C/code/model.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+            self.img2txt.load_state_dict(dicts['G12_state_dict'])
+            self.txt2img.load_state_dict(dicts['G21_state_dict'])
+            self.D_img.load_state_dict(dicts['D1_state_dict'])
+            self.D_txt.load_state_dict(dicts['D2_state_dict'])
+            self.optimizer_G.load_state_dict(dicts['optimizer_G'])
+            self.optimizer_D.load_state_dict(dicts['optimizer_D'])
+            # self.scheduler.load_state_dict(dicts['scheduler'])
+        else:
+            self.logger.error("> No checkpoint found at '{}'".format(cptpath))
+    def set_writer(self):
```
Comment (merelkuijs) [inline DM2C/code/model.py:354]: Put at least one (or better yet, two) blank lines between separate function and class definitions.

File: DM2C/code/utils.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        self.data = self.prepare_h5ad(h5ad_path)
+    def prepare_h5ad(self, h5ad_path):
+        data = ad.read_h5ad(h5ad_path)
+        data.obs_names_make_unique()
+        return data
+
+class img_Dataset(Dataset):
+    def __init__(self, img_path):
+        names = os.listdir(img_path)
+        self.imgs = [img_path+name for name in names]
```
Comment (merelkuijs) [inline DM2C/code/utils.py]: I prefer to use `os.path.join` instead of + when concatenating strings, if these strings represent paths.

File: DM2C/code/utils.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+    def __len__(self):
+        return len(self.imgs)
+
+class Custom_Dataloader():
+    def __init__(self, dataset, modal, batch_size=32, shuffle=True):
+        self.dataset, self.modal, self.batch_size, self.shuffle = dataset, modal, batch_size, shuffle
+    def __iter__(self):
+        indices = np.arange(len(self.dataset))
+        if self.shuffle:
+            indices = np.random.permutation(len(self.dataset))
```
Comment (merelkuijs) [inline DM2C/code/utils.py]: I think we need to be careful here. If we have many more images than gene expression vectors (or the other way around), we'll get very imbalanced batches. Even if our data is balanced, we might end up with imbalanced batches every once in a while simply by chance. Given a `batch_size`, it might be better to randomly pick `batch_size / 2` images and `batch_size / 2` cells.

File: DM2C/code/utils.py [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+        batches = []
```
Comment (merelkuijs) [inline DM2C/code/utils.py]: Do we use `batches`? I don't think so. Remove it to save memory.

File: DM2C/code/utils.py (hunk 13 above)
Comment (merelkuijs) [reply DM2C/code/utils.py]: I think the fastest way to implement it is to shuffle the image and cell indices separately, then pick `batch_size / 2` items from each until you run out of items for one or both of the modalities.

File: DM2C/code/model.py [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+        # we need to embed the h5ad data first so we can input them into the dataloader
+        h5ad_embed = torch.stack([embed for embed in self.cell_plm.calc_embed(h5ad_dataset.data)])
+        # same with images -- embed first
+        img_loader = DataLoader(dataset = img_dataset, # image loader so that not all images are read into memory for embedding calculation
+                                batch_size= self.args.batch_size,
+                                shuffle = True)
+        img_embed=[]
+        for load in img_loader:
+            img_embed.extend(self.vis_trans.calc_embed(load))
+        img_embed = torch.stack(img_embed)
```
Comment (merelkuijs) [inline DM2C/code/model.py:264]: Are you sure your code works as expected? Note that `torch.stack` creates a new dimension (the batch dimension), but the tensors returned by `self.vis_trans.calc_embed` have already been stacked! What is the shape of img_embed? It has to be (n_images, hidden_size).

File: DM2C/code/model.py [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+            self.save_cpt(epoch)
+
+    def _build_masterpraktikum_dataloader(self):
+        kwargs = {'num_workers': self.args.n_cpu, 'pin_memory': True}
+
+        h5ad_dataset = h5ad_Dataset(self.args.h5ad_data)
+        img_dataset = img_Dataset(self.args.img_path)
+
+        # we need to embed the h5ad data first so we can input them into the dataloader
+        h5ad_embed = torch.stack([embed for embed in self.cell_plm.calc_embed(h5ad_dataset.data)])
```
Comment (merelkuijs) [inline DM2C/code/model.py]: Doesn't `self.cell_plm.calc_embed` return a tensor shaped like (n_cells, hidden_size)? You do not need to create the list and then call `torch.stack`: it returns the same output as `self.cell_plm.calc_embed`, but takes longer.

File: DM2C/code/model.py (hunk 4 above)
Comment (RLuknarova) [reply DM2C/code/model.py:63]: it is outside of the masterpraktikum_ssl_folder in a folder where I did the CellPLM task. I wasnt sure if I should include that in my branch too

File: DM2C/code/model.py (hunk 15 above)
Comment (RLuknarova) [reply DM2C/code/model.py:264]: it works but there may be a better way to do it. maybe with concatenating
